stytch/src/main/kotlin/com/stytch/java/common/PolicyCache.kt

```diff
@@ -0,0 +1,54 @@
+package com.stytch.java.common
```

Comment:
Written by hand, very likely to not be idiomatic Kotlin.

Reply:
So I took a pass at cleaning this up a little, but left some questions in comments:

```
package com.stytch.java.common

import com.stytch.java.b2b.api.rbac.RBAC
import com.stytch.java.b2b.models.rbac.Policy
import com.stytch.java.b2b.models.rbac.PolicyRequest
import com.stytch.java.b2b.models.sessions.AuthorizationCheck
import java.time.Duration
import java.time.Instant

public class TenancyException(subjectOrgId: String, authCheckOrgId: String) :
    RuntimeException("Subject organizationId $subjectOrgId does not match authZ request organizationId $authCheckOrgId")

public class PermissionException(authorizationCheck: AuthorizationCheck) :
    RuntimeException("Permission denied for request $authorizationCheck")

internal class PolicyCache(
    private val client: RBAC,
) {
    private var cachedPolicy: Policy? = null
    private var policyLastUpdate: Instant? = null

    private fun getPolicy(invalidate: Boolean = false): Policy {
        val isMissing = cachedPolicy == null || policyLastUpdate == null
        val isStale = Duration.between(policyLastUpdate, Instant.now()).seconds > 300
        if (invalidate || isMissing || isStale) {
            when (val result = client.policyCompletable(PolicyRequest()).get()) {
                is StytchResult.Success -> {
                    cachedPolicy = result.value.policy
                    policyLastUpdate = Instant.now()
                }
                else -> {}
            }
        }
        return cachedPolicy ?: throw Exception("Error fetching the policy")
    }

    // Where does this get called from, I can't find any reference to it in the codebase
    fun performAuthorizationCheck(
        subjectRoles: List<String>,
        subjectOrgId: String,
        authorizationCheck: AuthorizationCheck,
    ) {
        if (authorizationCheck.organizationId != subjectOrgId) {
            throw TenancyException(subjectOrgId, authorizationCheck.organizationId)
        }
        val policy = getPolicy()
        val hasMatchingActionAndResource = policy.roles
            .filter { it.roleId in subjectRoles }
            .flatMap { it.permissions }
            .filter {
                val hasMatchingAction = it.actions.contains("*") || it.actions.contains(action) // where does action come from? It's undefined
                val hasMatchingResource = it.resourceId == authorizationCheck.resourceId
                return@filter hasMatchingAction && hasMatchingResource
            }
            .isNotEmpty()
        hasMatchingActionAndResource && return
        throw PermissionException(authorizationCheck)
    }
}
```

The above _should_ work, but would love to help write some unit tests around it :D